colossalai/inference/core/request_handler.py

```diff
@@ -1,48 +1,210 @@
+from colossalai.inference.kv_cache import KVCacheManager
+from colossalai.inference.logit_processors import logit_processor
+from colossalai.inference.sampler import *
+from colossalai.inference.struct import BatchInfo, RequsetStatus, Sequence
+
+
+class RunningList:
+    """
+    RunningList is an structure for recording the running sequences, contains prefill and decoding list.
+    Prefilling samples will be hold until the actual ratio of prefill samples versus decoding samples exceeds ratio.
+
+    Args:
+        prefill_ratio: float
+        decoding/prefill: list that contains prefill or decoding samples
+    """
+
+    def __init__(self, prefill_ratio):
+        self.prefill_ratio = prefill_ratio
+        self.decoding = []
+        self.prefill = []
+
+    def append(self, seq: Sequence):
+        # add seq to prefilling list first.
+        self.prefill.append(seq)
+
+    def find_seq(self, request_id):
+        for seq in self.decoding:
+            if request_id == seq.request_id:
+                return seq
+        for seq in self.prefill:
+            if request_id == seq.request_id:
+                return seq
+        return None
+
+    def remove(self, seq):
+        try:
+            self.decoding.remove(seq)
+        except:
+            self.prefill.remove(seq)
+
+    def ready_for_prefill(self):
+        if not self.decoding:
+            return len(self.prefill) > 0
+        return len(self.prefill) / len(self.decoding) >= self.ratio
+
+    def is_empty(self):
+        return not self.decoding and not self.prefill
+
 
 class RequestHandler:
     """
     RequestHandler is the core for handling existing requests and updating current batch.
     During generation process, we call schedule function each iteration to update current batch.
 
     Args:
-        cache_config: Configuration for initialize and manage kv cache.
+       inference_config: Configuration for initialize and manage kv cache.
     """
 
-    def __init__(self, cache_config) -> None:
-        self.cache_config = cache_config
-        self._init_cache()
-        self.waiting_list: List["Reqseq"] = []
-        self.running_list: List["Reqseq"] = []
+    def __init__(self, inference_config: InferenceConfig, model_config: PretrainedConfig) -> None:
+        self.inference_config = inference_config
+        self._init_cache(model_config)
 
-    def _init_cache(self):
-        """
-        Initialize the cache manager with cache config.
-        """
+        self.running_list: RunningList = RunningList(inference_config.prefill_ratio)
+        self.waiting_list: List[List] = [[], [], []]
+        self.done_list: List[Sequence] = []
+        self.running_batch = BatchInfo(is_prompts=False)
+        self.prefill_batch = BatchInfo(is_prompts=True)
+
+    def _init_cache(self, model_config):
+        self.cache_manager = KVCacheManager(self.inference_config, model_config)
+
+    def _has_waiting(self) -> bool:
+        return any(lst for lst in self.waiting_list)
```

Comment:
Do you want to determine whether waiting_list is empty? Why not just use len(waiting_list) == 0?

Reply:
`waiting_list` actually has 3 empty lists which determines the priority when scheduling, I think it is an easy way to decide which prompt has a higher priority 
![image](https://github.com/hpcaitech/ColossalAI/assets/72591262/26d3fd00-7eb1-4890-976a-dd5e5051ef5e)